File: C++/crow/serialization_handlers.cpp
#include <array>
#include <string>
#include <crow/http_response.h>
#include <glaze/glaze.hpp>
#include "serialization_handlers.h"

struct StaticUser {
    uint8_t id{};
    std::string_view username;
    std::string_view email;
    bool is_active;
    std::array<std::string_view, 2> roles;
};

template<>
struct glz::meta<StaticUser> {
    using T = StaticUser;
    static constexpr auto value = object(
        "id", &T::id,
        "username", &T::username,
        "email", &T::email,
        "is_active", &T::is_active,
        "roles", &T::roles
    );
};


crow::response handle_user_serialization() {
    try {
        StaticUser user = {
            1,
            "JohnDoe",
            "johndoe@gmail.com",
            true,
            {"user", "admin"}
        };
        crow::response res{crow::status::OK, glz::write_json(user).value()};
        res.set_header("Content-Type", "application/json");
        return res;
    } catch ([[maybe_unused]] const std::exception& e) {
        return {500, "Internal Server Error"};
    }
}

File: JavaScript-TypeScript/bun/src/serializationHandlers.ts
interface User {
  id: number
  username: string
  email: string
  isActive: boolean
  roles: string[]

  // Old class behavior
  // constructor(id: number, username: string, email: string, isActive: boolean, roles: string[]) {
  //   this.id = id
  //   this.username = username
  //   this.email = email
  //   this.isActive = isActive
  //   this.roles = roles
  // }
}


export function handleUserSerialization() {
  let user: User = {
    id: 1,
    username: "JohnDoe",
    email: "johndoe@gmail.com",
    isActive: true,
    roles: ["user", "admin"]
  }
  return Response.json(user)
}